### PR TITLE
adds flag to support skipping processing of street segments...

### PR DESCRIPTION
**Pull request by ddellacosta**
...and inserting after the output XML is generated.

https://www.pivotaltracker.com/story/show/156324653

There is still some work to be done here, but this is ready for a preliminary review.

A quick summary of how this is supposed to work:

- prep the 5.1 feed if it is in XML format by converting it to CSV using the [`vip-feed-format-converter`](https://github.com/votinginfoproject/vip-feed-format-converter)
- using the upload script, add the `:post-process-street-segments?` flag set to `true` to the SQS message args (doc [here](https://github.com/votinginfoproject/wiki/blob/master/data-processor/upload_script.md)).

That's it. What this will do is bypass loading of the `street_segments.txt` CSV file into the database, and at the end after an XML output file has been produced, will insert the street segments one-by-one, doing a very basic validation on each one (checks precinct id against what we've stored, and ensures the required fields exist for that street segment entry).

Testing has shown this speeds up the OH feed, for example, to about ~24 minutes vs. the ~12 hours it was taking.

TODO: more docstrings

### CODE REVIEW

**Comment by sduckett** [comment]
@ddellacosta Is this PR ready to review, or is there more coming?

**Comment by ddellacosta** [comment]
@sduckett 
I may add more docstrings, but it's been ready for a while--I was hoping to get a preliminary review in any case, as I stated in my initial comments.

File: dev-src/dev/core.clj [hunk 1]
```diff
@@ -1,25 +1,23 @@
...
-            [vip.data-processor.db.postgres :as psql]
             [vip.data-processor.s3 :refer [zip-filename]]
-            [squishy.data-readers]))
+            [vip.data-processor.validation.transforms :as t]
+            [vip.data-processor.validation.zip :as zip]))
 
+;; modify this to add flags to processing
+(defn set-context
+  [ctx]
+  (assoc ctx :post-process-street-segments? true))
```
Comment (tank157) [inline dev-src/dev/core.clj]: For clarification, `:post-process-street-segments?` set to true will skip the usual street segment load/validation processing, but add the street segments back to the output?

File: src/vip/data_processor/output/street_segments.clj [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+         read-one-line
+         (map #(str/replace % #"\W" ""))
+         (map (comp keyword str/lower-case)))))
+
+(defn street-segments
+  [ctx event-factory writer street-segment-file precinct-ids]
+  (with-open [in-file (util/bom-safe-reader street-segment-file :encoding "UTF-8")]
+    (let [header-row (header-row in-file)]
+      (doseq [line (csv/read-csv in-file)]
+        (let [line' (zipmap header-row line)]
```
Comment (tank157) [inline src/vip/data_processor/output/street_segments.clj]: small suggestion, couldn't you do `line'` in the `doseq` binding? I'm not 100% sure you can, but it seems like you can. Actually, I looked it up, and yeah, you can use a `:let [line' (zipmap header-row line)]`, but i figure at this point it's almost equivalent and harder to read, so maybe ignore this

File: src/vip/data_processor/output/street_segments.clj [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      (while (.hasNext reader)
+        (let [event (.nextEvent reader)]
+          (when (and (.isEndElement event)
+                     (= (str (.getName (.asEndElement event))) "VipObject"))
+            (street-segments ctx event-fact writer ss-file precinct-ids))
+          (.add writer event)))
+      (.close reader)
+      (.close writer)
+      (-> (assoc ctx :xml-output-file (.toPath tmpfile))
+          ;; Not sure if we care about this, we may want to just throw it away
```
Comment (sduckett) [inline src/vip/data_processor/output/street_segments.clj]: Agreed, throw it away.

File: src/vip/data_processor/validation/csv.clj [hunk 4]
```diff
@@ -18,28 +18,31 @@
...
-(defn good-filename? [data-specs file]
-  (let [filename (str/lower-case (file-name file))]
-    (contains? (csv-filenames data-specs) filename)))
-
-(defn-traced remove-bad-filenames [ctx]
-  (let [input (:input ctx)
-        {good-files true
-         bad-files false} (group-by (partial good-filename? (:data-specs ctx)) input)]
+(defn-traced remove-bad-filenames
+  "Filters out any files in `input` from our files to process based on
```
Comment (sduckett) [inline src/vip/data_processor/validation/csv.clj]: I might say _"Removes any files in..."_ for harmony with the function name.

File: src/vip/data_processor/output/street_segments.clj (hunk 2 above)
Comment (ddellacosta) [reply src/vip/data_processor/output/street_segments.clj]: Yeah, can't do it on the left side at least. I'll give the `:let` thing a try and see...could maybe be nicer, dunno.

**Pull request merged**